src/order/height.lean

```diff
@@ -0,0 +1,276 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+
+import data.list.chain
+import data.nat.lattice
+import algebra.order.monoid
+
+/-!
+
+# Maximal length of chains
+
+This file contains lemmas to work with the maximal length of strictly ascending finite
+sequences (chains) in a partially ordered set.
+
+## Main definition
+
+- `sup_chain_length α`: The maximal length of chains in `α`.
+- `chain_height a`: The maximal length of chains in `α` that start from `a : α`
+  (not including `a` itself).
+
+They are both valued in `enat`, and equal `⊤ : enat` when the lengths are unbounded.
+
+## Main results
+
+- `exists_chain_of_le_length`: For each `n : ℕ` such that `n ≤ chain_height a`, there exists a
+  chain of length `n` starting from `a`.
+- `chain_height_eq_one_iff`: An element has height zero iff it is minimal.
+- `chain_height_le_of_lt`: If `a < b`, then `chain_height b + 1 ≤ chain_height a`. This implies that
+  `chain_height` is decreasing (`chain_height_antitone`).
+- `sup_chain_length_eq_sup_chain_height`: `sup_chain_length` is the supremum of all the heights.
+- `sup_chain_length_dual`: The maximal length of a poset and is dual is the same.
+- `chain_height_add_chain_height_dual_le`: The sum of the height of the same element in `α` and
+  `αᵒᵈ` is bounded by `sup_chain_length`.
+
+-/
+
+universes u v
+
+namespace list
+
+variables (α : Type u) [partial_order α]
```

Comment:
You don't need a `partial_order` instance throughout the whole file. The definitions make sense with nothing but `has_lt`, and surely some of the theorems work on `preorder`s too?

Reply:
I restated everything in terms of `preorder`s. But I don't think there is there is any interesting type that has a `has_lt` instance but is not a `preorder`?

Reply:
I doubt there is one. But if the extra generality doesn't hurt, then I still believe you should restate the applicable results in terms of that.

Also, a definition or theorem with only a `has_lt` hypothesis is a good indicator that the result is only notational, and can thus be potentially generalized to other relations if needed.